jupyter/d2l-java/chapter_convolutional-modern/densenet.ipynb

```diff
@@ -0,0 +1,606 @@
+    "import ai.djl.training.optimizer.learningrate.LearningRateTracker;\n",
+    "import ai.djl.translate.Pipeline;\n",
+    "import ai.djl.util.PairList;\n",
+    "\n",
+    "import java.io.IOException;\n",
+    "import java.util.Arrays;\n",
+    "import java.util.HashMap;\n",
+    "import java.util.Map;\n",
+    "\n",
+    "import tech.tablesaw.api.*;\n",
+    "import tech.tablesaw.plotly.api.*;\n",
+    "import tech.tablesaw.plotly.components.*;\n",
+    "import tech.tablesaw.plotly.Plot;\n",
+    "import tech.tablesaw.plotly.components.Figure;\n",
+    "import org.apache.commons.lang3.ArrayUtils;"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "attributes": {
+     "classes": [],
+     "id": "",
+     "n": "1"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "public SequentialBlock convBlock(int numChannels) {\n",
+    "\n",
+    "        SequentialBlock block = new SequentialBlock()\n",
+    "                .add(BatchNorm.builder().build())\n",
+    "                .add(Activation::relu)\n",
+    "                .add(Conv2d.builder()\n",
+    "                        .setFilters(numChannels)\n",
+    "                        .setKernelShape(new Shape(3, 3))\n",
+    "                        .optPadding(new Shape(1, 1))\n",
+    "                        .optStride(new Shape(1, 1))\n",
+    "                        .build()\n",
+    "                );\n",
+    "\n",
+    "        return block;\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "A dense block consists of multiple `convBlock` units, each using the same number of output channels. In the forward computation, however, we concatenate the input and output of each block on the channel dimension."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "attributes": {
+     "classes": [],
+     "id": "",
+     "n": "2"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "class DenseBlock extends AbstractBlock {\n",
+    "\n",
+    "    private static final byte VERSION = 2;\n",
+    "    public SequentialBlock net = new SequentialBlock();\n",
+    "\n",
+    "    public DenseBlock(int numConvs, int numChannels) {\n",
+    "        super(VERSION);\n",
+    "        for (int i = 0; i < numConvs; i++) {\n",
+    "            this.net.add(\n",
+    "                    addChildBlock(\"denseBlock\" + i, convBlock(numChannels))\n",
+    "            );\n",
+    "        }\n",
+    "    }\n",
+    "\n",
+    "    @Override\n",
+    "    public String toString() {\n",
+    "        return \"DenseBlock()\";\n",
+    "    }\n",
+    "\n",
+    "    @Override\n",
+    "    public NDList forward(ParameterStore parameterStore, NDList X, boolean training, PairList<String, Object> pairList) {\n",
+    "\n",
+    "        NDArray Y;\n",
+    "        for (Block block : this.net.getChildren().values()) {\n",
```

Comment:
What does this line do? Not sure if we can replace by using Parallel block since it seemed to parallel run all elements

Reply:
![image](https://user-images.githubusercontent.com/11890922/88123247-570d5c80-cb7f-11ea-871b-343d4aa0cb6b.png)
After checking, it seemed you can use a parallel block to represent each layer to do the execution work.

Reply:
The for loop is doing the following thing (+ means concat)
```
iter 0: f1(X) + X
iter 1: f2(f1(X) + X) + X
iter 2: f3(f2(f1(X) + X) + X)
...
```

Reply:
I would say there is no easier solution than the current one